examples/MicroserviceExample/WorkerService/MessageProcessor.cs

```diff
@@ -0,0 +1,86 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using OpenTelemetry;
+using OpenTelemetry.Context.Propagation;
+using OpenTelemetry.Trace;
+using RabbitMQ.Client;
+using RabbitMQ.Client.Events;
+
+namespace WorkerService
+{
+    public class MessageProcessor : IDisposable
+    {
+        private static readonly ActivitySource ActivitySource = new ActivitySource(nameof(MessageProcessor));
+        private static readonly ITextFormat TextFormat = new TraceContextFormat();
+
+        private readonly TracerProvider tracerProvider;
+
+        public MessageProcessor()
+        {
+            this.tracerProvider = Sdk.CreateTracerProvider((builder) =>
+            {
+                builder
+                    .AddActivitySource(nameof(MessageProcessor))
+                    .UseZipkinExporter(b =>
+                    {
+                        var zipkinHostName = Environment.GetEnvironmentVariable("ZIPKIN_HOSTNAME") ?? "localhost";
+                        b.ServiceName = nameof(WorkerService);
+                        b.Endpoint = new Uri($"http://{zipkinHostName}:9411/api/v2/spans");
+                    });
+            });
+        }
+
+        public void Dispose()
+        {
+            this.tracerProvider.Dispose();
+        }
+
+        public async Task ProcessMessage(BasicDeliverEventArgs ea)
+        {
+            // Extract the ActivityContext of the upstream parent from the message headers.
+            var parentContext = TextFormat.Extract(ea.BasicProperties, ExtractTraceContextFromBasicProperties);
+
+            // Start an activity with a name following the semantic convention of the OpenTelemetry messaging specification.
+            // https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/messaging.md#span-name
+            var activityName = $"{ea.RoutingKey} receive";
+            using (var activity = ActivitySource.StartActivity(activityName, ActivityKind.Server, parentContext))
+            {
+                try
+                {
+                    var message = Encoding.UTF8.GetString(ea.Body.Span);
+
+                    activity.AddTag("message", message);
```

Comment:
There are some other semantic tags we could demo here, like `messaging.system`, `messaging.destination`, `messaging.destination_kind`. Might be useful to show a mix of semantic tags and custom ones?

Reply:
Yes all good suggestions. I also noted that the spec calls out some [special considerations](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/messaging.md#rabbitmq) for RabbitMQ.

Reply:
Doh. I haven't noticed that special RabbitMQ rule before, I'm going to have to fix some stuff IRL 😆